SDL2 surface tests fail

```
loading test.surface_test
FE.E.....E...........F...................s.......s.F..............
======================================================================
ERROR: test_blit_blend (test.surface_test.SurfaceBlendTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\surface_test.py", line 1940, in test_blit_blend
    [masks[1], masks[2], masks[0], masks[3]])
ValueError: Invalid mask values

======================================================================
ERROR: test_blit_blend_rgba (test.surface_test.SurfaceBlendTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\surface_test.py", line 2031, in test_blit_blend_rgba
    (masks[1], masks[2], masks[3], masks[0]))
ValueError: Invalid mask values

======================================================================
ERROR: test_array_interface_masks (test.surface_test.SurfaceGetBufferTest)
Test non-default color byte orders on 3D views
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\surface_test.py", line 1518, in test_array_interface_masks
    pygame.Surface(sz, 0, 24, masks).get_view, '3')
ValueError: Invalid mask values

======================================================================
FAIL: test_GET_PIXELVALS (test.surface_test.SurfaceBlendTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\surface_test.py", line 2098, in test_GET_PIXELVALS
    self.assertEqual(dst.get_at((0, 0)), (0, 0, 0, 255))
AssertionError: (0, 0, 0, 128) != (0, 0, 0, 255)

======================================================================
FAIL: test_overlap_check (test.surface_test.SurfaceSelfBlitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\surface_test.py", line 2241, in test_overlap_check
    self.assertEqual(surf.get_at(test_posn), rectc_right)
AssertionError: (128, 64, 32, 255) != (255, 255, 255, 255)

======================================================================
FAIL: test_image_convert_bug_131 (test.surface_test.SurfaceTypeTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\surface_test.py", line 674, in test_image_convert_bug_131
    self.assertRaises(pygame.error, im2.convert, 8)
AssertionError: error not raised by convert

----------------------------------------------------------------------
Ran 66 tests in 5.537s

FAILED (failures=3, errors=3)
```

`test_blanket_alpha` fails again:

```
======================================================================
FAIL: test_blanket_alpha (test.surface_test.SurfaceSelfBlitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\surface_test.py", line 2323, in test_blanket_alpha
    self._assert_same(surf, comp)
  File "C:\Users\David\Documents\dev\pygame\test\surface_test.py", line 2243, in _assert_same
    a.get_bitsize())))
AssertionError: (255, 0, 0, 255) != (85, 146, 0, 255) : (255, 0, 0, 255) != (85, 146, 0, 255), bpp: 8
```

--- a/src_c/alphablit.c
+++ b/src_c/alphablit.c
@@ -45,6 +45,8 @@ typedef struct
     Uint8            src_blanket_alpha;
     int              src_has_colorkey;
     Uint32           src_colorkey;
+    SDL_BlendMode    src_blend;
+    SDL_BlendMode    dst_blend;
 #endif /* IS_SDLv2 */
 } SDL_BlitInfo;
 
@@ -73,31 +75,6 @@ extern int  SDL_RLESurface (SDL_Surface * surface);
 extern void SDL_UnRLESurface (SDL_Surface * surface, int recode);
 
 
-#if IS_SDLv2
-static int
-_PgSurface_SrcAlpha(SDL_Surface *surf)
-{
-    if (SDL_ISPIXELFORMAT_ALPHA(surf->format->format)) {
-        SDL_BlendMode mode;
-        if (SDL_GetSurfaceBlendMode(surf, &mode) < 0) {
-            return -1;
-        }
-        if (mode == SDL_BLENDMODE_BLEND)
-            return 1;
-    }
-    else {
-        Uint8 color = SDL_ALPHA_OPAQUE;
-        if (SDL_GetSurfaceAlphaMod(surf, &color) != 0) {
-            return -1;
-        }
-        if (color != SDL_ALPHA_OPAQUE)
-            return 1;
-    }
-    return 0;
-}
-#endif /* IS_SDLv2 */
-
-
 
 static int
 SoftBlitPyGame (SDL_Surface * src, SDL_Rect * srcrect, SDL_Surface * dst,
@@ -106,9 +83,6 @@ SoftBlitPyGame (SDL_Surface * src, SDL_Rect * srcrect, SDL_Surface * dst,
     int okay;
     int src_locked;
     int dst_locked;
-#if IS_SDLv2
-    int isalpha;
-#endif
 
     /* Everything is okay at the beginning...  */
     okay = 1;
@@ -169,6 +143,11 @@ SoftBlitPyGame (SDL_Surface * src, SDL_Rect * srcrect, SDL_Surface * dst,
 #else /* IS_SDLv2 */
         SDL_GetSurfaceAlphaMod (src, &info.src_blanket_alpha);
         info.src_has_colorkey = SDL_GetColorKey (src, &info.src_colorkey) == 0;
+        if (SDL_GetSurfaceBlendMode(src, &info.src_blend) ||
+            SDL_GetSurfaceBlendMode(dst, &info.dst_blend)) {
+            okay = 0;
+            goto LEAVE;
+        }
 #endif /* IS_SDLv2 */
 
         if (info.d_pixels > info.s_pixels)
@@ -211,11 +190,8 @@ SoftBlitPyGame (SDL_Surface * src, SDL_Rect * srcrect, SDL_Surface * dst,
                 alphablit_solid (&info);
             break;
 #else /* IS_SDLv2 */
-            isalpha = _PgSurface_SrcAlpha(src);
-            if (isalpha == 1) {
+            if (info.src_blend != SDL_BLENDMODE_NONE && src->format->Amask) {
                 alphablit_alpha (&info);
-            } else if (isalpha == -1) {
-                okay = 0;
             } else if (info.src_has_colorkey) {
                 alphablit_colorkey (&info);
             } else {
@@ -288,6 +264,9 @@ SoftBlitPyGame (SDL_Surface * src, SDL_Rect * srcrect, SDL_Surface * dst,
         }
         }
     }
+
+LEAVE:
+
     /* We need to unlock the surfaces if they're locked */
     if (dst_locked)
         SDL_UnlockSurface (dst);
@@ -330,8 +309,8 @@ blit_blend_rgba_add (SDL_BlitInfo * info)
     int             srcppa = (info->src_flags & SDL_SRCALPHA && srcfmt->Amask);
     int             dstppa = (info->dst_flags & SDL_SRCALPHA && dstfmt->Amask);
 #else /* IS_SDLv2 */
-    int             srcppa = SDL_ISPIXELFORMAT_ALPHA (srcfmt->format);
-    int             dstppa = SDL_ISPIXELFORMAT_ALPHA (dstfmt->format);
+    int             srcppa = info->src_blend != SDL_BLENDMODE_NONE && srcfmt->Amask;
+    int             dstppa = info->dst_blend != SDL_BLENDMODE_NONE && dstfmt->Amask;
 #endif /* IS_SDLv2 */
 
     if (!dstppa)
@@ -353,7 +332,7 @@ blit_blend_rgba_add (SDL_BlitInfo * info)
         srcfmt->Gmask == dstfmt->Gmask &&
         srcfmt->Bmask == dstfmt->Bmask &&
         srcfmt->Amask == dstfmt->Amask &&
-        srcppa)
+        info->src_blend != SDL_BLENDMODE_NONE)
 #endif /* IS_SDLv2 */
     {
         int incr = srcpxskip > 0 ? 1 : -1;
@@ -484,8 +463,8 @@ blit_blend_rgba_sub (SDL_BlitInfo * info)
     int             srcppa = (info->src_flags & SDL_SRCALPHA && srcfmt->Amask);
     int             dstppa = (info->dst_flags & SDL_SRCALPHA && dstfmt->Amask);
 #else /* IS_SDLv2 */
-    int             srcppa = SDL_ISPIXELFORMAT_ALPHA (srcfmt->format);
-    int             dstppa = SDL_ISPIXELFORMAT_ALPHA (dstfmt->format);
+    int             srcppa = info->src_blend != SDL_BLENDMODE_NONE && srcfmt->Amask;
+    int             dstppa = info->dst_blend != SDL_BLENDMODE_NONE && dstfmt->Amask;
 #endif /* IS_SDLv2 */
 
     if (!dstppa)
@@ -507,7 +486,7 @@ blit_blend_rgba_sub (SDL_BlitInfo * info)
         srcfmt->Gmask == dstfmt->Gmask &&
         srcfmt->Bmask == dstfmt->Bmask &&
         srcfmt->Amask == dstfmt->Amask &&
-        srcppa)
+        info->src_blend != SDL_BLENDMODE_NONE)
 #endif /* IS_SDLv2 */
     {
         int incr = srcpxskip > 0 ? 1 : -1;
@@ -638,8 +617,8 @@ blit_blend_rgba_mul (SDL_BlitInfo * info)
     int             srcppa = (info->src_flags & SDL_SRCALPHA && srcfmt->Amask);
     int             dstppa = (info->dst_flags & SDL_SRCALPHA && dstfmt->Amask);
 #else /* IS_SDLv2 */
-    int             srcppa = SDL_ISPIXELFORMAT_ALPHA (srcfmt->format);
-    int             dstppa = SDL_ISPIXELFORMAT_ALPHA (dstfmt->format);
+    int             srcppa = info->src_blend != SDL_BLENDMODE_NONE && srcfmt->Amask;
+    int             dstppa = info->dst_blend != SDL_BLENDMODE_NONE && dstfmt->Amask;
 #endif /* IS_SDLv2 */
 
     if (!dstppa)
@@ -656,7 +635,7 @@ blit_blend_rgba_mul (SDL_BlitInfo * info)
 #if IS_SDLv1
         info->src_flags & SDL_SRCALPHA)
 #else /* IS_SDLv2 */
-        srcppa)
+        info->src_blend != SDL_BLENDMODE_NONE)
 #endif /* IS_SDLv2 */
     {
         int incr = srcpxskip > 0 ? 1 : -1;
@@ -786,8 +765,8 @@ blit_blend_rgba_min (SDL_BlitInfo * info)
     int             srcppa = (info->src_flags & SDL_SRCALPHA && srcfmt->Amask);
     int             dstppa = (info->dst_flags & SDL_SRCALPHA && dstfmt->Amask);
 #else /* IS_SDLv2 */
-    int             srcppa = SDL_ISPIXELFORMAT_ALPHA (srcfmt->format);
-    int             dstppa = SDL_ISPIXELFORMAT_ALPHA (dstfmt->format);
+    int             srcppa = info->src_blend != SDL_BLENDMODE_NONE && srcfmt->Amask;
+    int             dstppa = info->dst_blend != SDL_BLENDMODE_NONE && dstfmt->Amask;
 #endif /* IS_SDLv2 */
 
     if (!dstppa)
@@ -804,7 +783,7 @@ blit_blend_rgba_min (SDL_BlitInfo * info)
 #if IS_SDLv1
         info->src_flags & SDL_SRCALPHA)
 #else /* IS_SDLv2 */
-        srcppa)
+        info->src_blend != SDL_BLENDMODE_NONE)
 #endif /* IS_SDLv2 */
     {
         int incr = srcpxskip > 0 ? 1 : -1;
@@ -934,8 +913,8 @@ blit_blend_rgba_max (SDL_BlitInfo * info)
     int             srcppa = (info->src_flags & SDL_SRCALPHA && srcfmt->Amask);
     int             dstppa = (info->dst_flags & SDL_SRCALPHA && dstfmt->Amask);
 #else /* IS_SDLv2 */
-    int             srcppa = SDL_ISPIXELFORMAT_ALPHA (srcfmt->format);
-    int             dstppa = SDL_ISPIXELFORMAT_ALPHA (dstfmt->format);
+    int             srcppa = info->src_blend != SDL_BLENDMODE_NONE && srcfmt->Amask;
+    int             dstppa = info->dst_blend != SDL_BLENDMODE_NONE && dstfmt->Amask;
 #endif /* IS_SDLv2 */
 
     if (!dstppa)
@@ -952,7 +931,7 @@ blit_blend_rgba_max (SDL_BlitInfo * info)
 #if IS_SDLv1
         info->src_flags & SDL_SRCALPHA)
 #else /* IS_SDLv2 */
-        srcppa)
+        info->src_blend != SDL_BLENDMODE_NONE)
 #endif /* IS_SDLv2 */
     {
         int incr = srcpxskip > 0 ? 1 : -1;
@@ -1081,7 +1060,10 @@ blit_blend_premultiplied (SDL_BlitInfo * info)
 #if IS_SDLv1
     int             srcppa = (info->src_flags & SDL_SRCALPHA && srcfmt->Amask);
     int             dstppa = (info->dst_flags & SDL_SRCALPHA && dstfmt->Amask);
-#endif /* IS_SDLv1 */
+#else /* IS_SDLv2 */
+    int             srcppa = info->src_blend != SDL_BLENDMODE_NONE && srcfmt->Amask;
+    int             dstppa = info->dst_blend != SDL_BLENDMODE_NONE && dstfmt->Amask;
+#endif /* IS_SDLv2 */
     int tmp;
 
     /*
@@ -1207,13 +1189,14 @@ blit_blend_add (SDL_BlitInfo * info)
     int             srcppa = (info->src_flags & SDL_SRCALPHA && srcfmt->Amask);
     int             dstppa = (info->dst_flags & SDL_SRCALPHA && dstfmt->Amask);
 #else /* IS_SDLv2 */
-    int             srcppa = SDL_ISPIXELFORMAT_ALPHA (srcfmt->format);
+    int             srcppa = info->src_blend != SDL_BLENDMODE_NONE && srcfmt->Amask;
+    int             dstppa = info->dst_blend != SDL_BLENDMODE_NONE && dstfmt->Amask;
 #endif /* IS_SDLv2 */
 
 #if IS_SDLv1
     if (srcbpp >= 3 && dstbpp >= 3 && !(info->src_flags & SDL_SRCALPHA))
 #else /* IS_SDLv2 */
-    if (srcbpp >= 3 && dstbpp >= 3 && !srcppa)
+    if (srcbpp >= 3 && dstbpp >= 3 && info->src_blend == SDL_BLENDMODE_NONE)
 #endif /* IS_SDLv2 */
     {
         size_t srcoffsetR, srcoffsetG, srcoffsetB;
@@ -1402,13 +1385,14 @@ blit_blend_sub (SDL_BlitInfo * info)
     int             srcppa = (info->src_flags & SDL_SRCALPHA && srcfmt->Amask);
     int             dstppa = (info->dst_flags & SDL_SRCALPHA && dstfmt->Amask);
 #else /* IS_SDLv2 */
-    int             srcppa = SDL_ISPIXELFORMAT_ALPHA (srcfmt->format);
+    int             srcppa = info->src_blend != SDL_BLENDMODE_NONE && srcfmt->Amask;
+    int             dstppa = info->dst_blend != SDL_BLENDMODE_NONE && dstfmt->Amask;
 #endif /* IS_SDLv2 */
 
 #if IS_SDLv1
     if (srcbpp >= 3 && dstbpp >= 3 && !(info->src_flags & SDL_SRCALPHA))
 #else /* IS_SDLv2 */
-    if (srcbpp >= 3 && dstbpp >= 3 && !srcppa)
+    if (srcbpp >= 3 && dstbpp >= 3 && info->src_blend == SDL_BLENDMODE_NONE)
 #endif /* IS_SDLv2 */
     {
         size_t srcoffsetR, srcoffsetG, srcoffsetB;
@@ -1597,13 +1581,14 @@ blit_blend_mul (SDL_BlitInfo * info)
     int             srcppa = (info->src_flags & SDL_SRCALPHA && srcfmt->Amask);
     int             dstppa = (info->dst_flags & SDL_SRCALPHA && dstfmt->Amask);
 #else /* IS_SDLv2 */
-    int             srcppa = SDL_ISPIXELFORMAT_ALPHA (srcfmt->format);
+    int             srcppa = info->src_blend != SDL_BLENDMODE_NONE && srcfmt->Amask;
+    int             dstppa = info->dst_blend != SDL_BLENDMODE_NONE && dstfmt->Amask;
 #endif /* IS_SDLv2 */
 
 #if IS_SDLv1
     if (srcbpp >= 3 && dstbpp >= 3 && !(info->src_flags & SDL_SRCALPHA))
 #else /* IS_SDLv2 */
-    if (srcbpp >= 3 && dstbpp >= 3 && !srcppa)
+    if (srcbpp >= 3 && dstbpp >= 3 && info->src_blend == SDL_BLENDMODE_NONE)
 #endif /* IS_SDLv2 */
     {
         size_t srcoffsetR, srcoffsetG, srcoffsetB;
@@ -1795,13 +1780,14 @@ blit_blend_min (SDL_BlitInfo * info)
     int             srcppa = (info->src_flags & SDL_SRCALPHA && srcfmt->Amask);
     int             dstppa = (info->dst_flags & SDL_SRCALPHA && dstfmt->Amask);
 #else /* IS_SDLv2 */
-    int             srcppa = SDL_ISPIXELFORMAT_ALPHA (srcfmt->format);
+    int             srcppa = info->src_blend != SDL_BLENDMODE_NONE && srcfmt->Amask;
+    int             dstppa = info->dst_blend != SDL_BLENDMODE_NONE && dstfmt->Amask;
 #endif /* IS_SDLv2 */
 
 #if IS_SDLv1
     if (srcbpp >= 3 && dstbpp >= 3 && !(info->src_flags & SDL_SRCALPHA))
 #else /* IS_SDLv2 */
-    if (srcbpp >= 3 && dstbpp >= 3 && !srcppa)
+    if (srcbpp >= 3 && dstbpp >= 3 && info->src_blend == SDL_BLENDMODE_NONE)
 #endif /* IS_SDLv2 */
     {
         size_t srcoffsetR, srcoffsetG, srcoffsetB;
@@ -1995,13 +1981,14 @@ blit_blend_max (SDL_BlitInfo * info)
     int             srcppa = (info->src_flags & SDL_SRCALPHA && srcfmt->Amask);
     int             dstppa = (info->dst_flags & SDL_SRCALPHA && dstfmt->Amask);
 #else /* IS_SDLv2 */
-    int             srcppa = SDL_ISPIXELFORMAT_ALPHA (srcfmt->format);
+    int             srcppa = info->src_blend != SDL_BLENDMODE_NONE && srcfmt->Amask;
+    int             dstppa = info->dst_blend != SDL_BLENDMODE_NONE && dstfmt->Amask;
 #endif /* IS_SDLv2 */
 
 #if IS_SDLv1
     if (srcbpp >= 3 && dstbpp >= 3 && !(info->src_flags & SDL_SRCALPHA))
 #else /* IS_SDLv2 */
-    if (srcbpp >= 3 && dstbpp >= 3 && !srcppa)
+    if (srcbpp >= 3 && dstbpp >= 3 && info->src_blend == SDL_BLENDMODE_NONE)
 #endif /* IS_SDLv2 */
     {
         size_t srcoffsetR, srcoffsetG, srcoffsetB;

--- a/src_c/surface.c
+++ b/src_c/surface.c
@@ -1514,7 +1514,7 @@ surf_set_alpha(PyObject *self, PyObject *args)
 #endif /* IS_SDLv2 */
     }
 #if IS_SDLv2
-    else if (SDL_ISPIXELFORMAT_ALPHA(surf->format->format)) {
+    else {
         if (SDL_SetSurfaceBlendMode(surf, SDL_BLENDMODE_NONE) != 0)
             return RAISE(pgExc_SDLError, SDL_GetError());
     }
@@ -1574,11 +1574,11 @@ surf_get_alpha(PyObject *self)
     if (SDL_GetSurfaceBlendMode(surf, &mode) != 0)
         return RAISE(pgExc_SDLError, SDL_GetError());
 
+    if (mode != SDL_BLENDMODE_BLEND)
+        Py_RETURN_NONE;
+
     if (SDL_GetSurfaceAlphaMod(surf, &alpha) != 0)
         return RAISE(pgExc_SDLError, SDL_GetError());
-
-    if (alpha == SDL_ALPHA_OPAQUE && mode != SDL_BLENDMODE_BLEND)
-        Py_RETURN_NONE;
 
     return PyInt_FromLong(alpha);
 #endif /* IS_SDLv2 */
@@ -2462,25 +2462,12 @@ surf_scroll(PyObject *self, PyObject *args, PyObject *keywds)
 static int
 _PgSurface_SrcAlpha(SDL_Surface *surf)
 {
-    if (SDL_ISPIXELFORMAT_ALPHA(surf->format->format)) {
-        SDL_BlendMode mode;
-        if (SDL_GetSurfaceBlendMode(surf, &mode) < 0) {
-            RAISE(pgExc_SDLError, SDL_GetError());
-            return -1;
-        }
-        if (mode == SDL_BLENDMODE_BLEND)
-            return 1;
+    SDL_BlendMode mode;
+    if (SDL_GetSurfaceBlendMode(surf, &mode) < 0) {
+        RAISE(pgExc_SDLError, SDL_GetError());
+        return -1;
     }
-    else {
-        Uint8 color = SDL_ALPHA_OPAQUE;
-        if (SDL_GetSurfaceAlphaMod(surf, &color) != 0) {
-            RAISE(pgExc_SDLError, SDL_GetError());
-            return -1;
-        }
-        if (color != SDL_ALPHA_OPAQUE)
-            return 1;
-    }
-    return 0;
+    return (mode != SDL_BLENDMODE_NONE);
 }
 #endif /* IS_SDLv2 */
 

--- a/src_c/surface.h
+++ b/src_c/surface.h
@@ -107,14 +107,17 @@
 #define SET_PIXELVAL(px, fmt, _dR, _dG, _dB, _dA) \
     *(px) = (Uint8) SDL_MapRGB(fmt, _dR, _dG, _dB)
 #else /* IS_SDLv2 */
-#define GET_PIXELVALS(_sR, _sG, _sB, _sA, px, fmt, ppa) \
-    SDL_GetRGBA(px, fmt, &(_sR), &(_sG), &(_sB), &(_sA))
+#define GET_PIXELVALS(_sR, _sG, _sB, _sA, px, fmt, ppa)   \
+    SDL_GetRGBA(px, fmt, &(_sR), &(_sG), &(_sB), &(_sA)); \
+    if (!ppa) {                                           \
+        _sA = 255;                                        \
+    }
 
 #define GET_PIXELVALS_1(sr, sg, sb, sa, _src, _fmt)    \
     sr = _fmt->palette->colors[*((Uint8 *) (_src))].r; \
     sg = _fmt->palette->colors[*((Uint8 *) (_src))].g; \
     sb = _fmt->palette->colors[*((Uint8 *) (_src))].b; \
-    sa = _fmt->palette->colors[*((Uint8 *) (_src))].a;
+    sa = 255;
 
 /* For 1 byte palette pixels */
 #define SET_PIXELVAL(px, fmt, _dR, _dG, _dB, _dA) \

--- a/src_c/surface_fill.c
+++ b/src_c/surface_fill.c
@@ -92,7 +92,12 @@ surface_fill_blend_add(SDL_Surface *surface, SDL_Rect *rect, Uint32 color)
     int result = -1;
 #if IS_SDLv1
     int ppa = (surface->flags & SDL_SRCALPHA && fmt->Amask);
-#endif /* IS_SDLv1 */
+#else /* IS_SDLv2 */
+    int ppa;
+    SDL_BlendMode mode;
+    SDL_GetSurfaceBlendMode(surface, &mode);
+    ppa = (fmt->Amask && mode != SDL_BLENDMODE_NONE);
+#endif /* IS_SDLv2 */
 
 #if IS_SDLv1
     pixels = (Uint8 *)surface->pixels + surface->offset +
@@ -181,7 +186,12 @@ surface_fill_blend_sub(SDL_Surface *surface, SDL_Rect *rect, Uint32 color)
     int result = -1;
 #if IS_SDLv1
     int ppa = (surface->flags & SDL_SRCALPHA && fmt->Amask);
-#endif /* IS_SDLv1 */
+#else /* IS_SDLv2 */
+    int ppa;
+    SDL_BlendMode mode;
+    SDL_GetSurfaceBlendMode(surface, &mode);
+    ppa = (fmt->Amask && mode != SDL_BLENDMODE_NONE);
+#endif /* IS_SDLv2 */
 
 #if IS_SDLv1
     pixels = (Uint8 *)surface->pixels + surface->offset +
@@ -269,7 +279,12 @@ surface_fill_blend_mult(SDL_Surface *surface, SDL_Rect *rect, Uint32 color)
     int result = -1;
 #if IS_SDLv1
     int ppa = (surface->flags & SDL_SRCALPHA && fmt->Amask);
-#endif /* IS_SDLv1 */
+#else /* IS_SDLv2 */
+    int ppa;
+    SDL_BlendMode mode;
+    SDL_GetSurfaceBlendMode(surface, &mode);
+    ppa = (fmt->Amask && mode != SDL_BLENDMODE_NONE);
+#endif /* IS_SDLv2 */
 
 #if IS_SDLv1
     pixels = (Uint8 *)surface->pixels + surface->offset +
@@ -357,7 +372,12 @@ surface_fill_blend_min(SDL_Surface *surface, SDL_Rect *rect, Uint32 color)
     int result = -1;
 #if IS_SDLv1
     int ppa = (surface->flags & SDL_SRCALPHA && fmt->Amask);
-#endif /* IS_SDLv1 */
+#else /* IS_SDLv2 */
+    int ppa;
+    SDL_BlendMode mode;
+    SDL_GetSurfaceBlendMode(surface, &mode);
+    ppa = (fmt->Amask && mode != SDL_BLENDMODE_NONE);
+#endif /* IS_SDLv2 */
 
 #if IS_SDLv1
     pixels = (Uint8 *)surface->pixels + surface->offset +
@@ -445,7 +465,12 @@ surface_fill_blend_max(SDL_Surface *surface, SDL_Rect *rect, Uint32 color)
     int result = -1;
 #if IS_SDLv1
     int ppa = (surface->flags & SDL_SRCALPHA && fmt->Amask);
-#endif /* IS_SDLv1 */
+#else /* IS_SDLv2 */
+    int ppa;
+    SDL_BlendMode mode;
+    SDL_GetSurfaceBlendMode(surface, &mode);
+    ppa = (fmt->Amask && mode != SDL_BLENDMODE_NONE);
+#endif /* IS_SDLv2 */
 
 #if IS_SDLv1
     pixels = (Uint8 *)surface->pixels + surface->offset +
@@ -536,8 +561,11 @@ surface_fill_blend_rgba_add(SDL_Surface *surface, SDL_Rect *rect, Uint32 color)
     int result = -1;
 #if IS_SDLv1
     int ppa = (surface->flags & SDL_SRCALPHA && fmt->Amask);
-#else  /* IS_SDLv2 */
-    int ppa = SDL_ISPIXELFORMAT_ALPHA(fmt->format);
+#else /* IS_SDLv2 */
+    int ppa;
+    SDL_BlendMode mode;
+    SDL_GetSurfaceBlendMode(surface, &mode);
+    ppa = (fmt->Amask && mode != SDL_BLENDMODE_NONE);
 #endif /* IS_SDLv2 */
 
     if (!ppa) {
@@ -610,8 +638,11 @@ surface_fill_blend_rgba_sub(SDL_Surface *surface, SDL_Rect *rect, Uint32 color)
     int result = -1;
 #if IS_SDLv1
     int ppa = (surface->flags & SDL_SRCALPHA && fmt->Amask);
-#else  /* IS_SDLv2 */
-    int ppa = SDL_ISPIXELFORMAT_ALPHA(fmt->format);
+#else /* IS_SDLv2 */
+    int ppa;
+    SDL_BlendMode mode;
+    SDL_GetSurfaceBlendMode(surface, &mode);
+    ppa = (fmt->Amask && mode != SDL_BLENDMODE_NONE);
 #endif /* IS_SDLv2 */
 
     if (!ppa) {
@@ -684,8 +715,11 @@ surface_fill_blend_rgba_mult(SDL_Surface *surface, SDL_Rect *rect,
     int result = -1;
 #if IS_SDLv1
     int ppa = (surface->flags & SDL_SRCALPHA && fmt->Amask);
-#else  /* IS_SDLv2 */
-    int ppa = SDL_ISPIXELFORMAT_ALPHA(fmt->format);
+#else /* IS_SDLv2 */
+    int ppa;
+    SDL_BlendMode mode;
+    SDL_GetSurfaceBlendMode(surface, &mode);
+    ppa = (fmt->Amask && mode != SDL_BLENDMODE_NONE);
 #endif /* IS_SDLv2 */
 
     if (!ppa) {
@@ -757,8 +791,11 @@ surface_fill_blend_rgba_min(SDL_Surface *surface, SDL_Rect *rect, Uint32 color)
     int result = -1;
 #if IS_SDLv1
     int ppa = (surface->flags & SDL_SRCALPHA && fmt->Amask);
-#else  /* IS_SDLv2 */
-    int ppa = SDL_ISPIXELFORMAT_ALPHA(fmt->format);
+#else /* IS_SDLv2 */
+    int ppa;
+    SDL_BlendMode mode;
+    SDL_GetSurfaceBlendMode(surface, &mode);
+    ppa = (fmt->Amask && mode != SDL_BLENDMODE_NONE);
 #endif /* IS_SDLv2 */
 
     if (!ppa) {
@@ -830,8 +867,11 @@ surface_fill_blend_rgba_max(SDL_Surface *surface, SDL_Rect *rect, Uint32 color)
     int result = -1;
 #if IS_SDLv1
     int ppa = (surface->flags & SDL_SRCALPHA && fmt->Amask);
-#else  /* IS_SDLv2 */
-    int ppa = SDL_ISPIXELFORMAT_ALPHA(fmt->format);
+#else /* IS_SDLv2 */
+    int ppa;
+    SDL_BlendMode mode;
+    SDL_GetSurfaceBlendMode(surface, &mode);
+    ppa = (fmt->Amask && mode != SDL_BLENDMODE_NONE);
 #endif /* IS_SDLv2 */
 
     if (!ppa) {

--- a/test/surface_test.py
+++ b/test/surface_test.py
@@ -643,11 +643,6 @@ class SurfaceTypeTest(AssertRaisesRegexMixin, unittest.TestCase):
         self.assertEqual(repr(im3), '<Surface(24x24x8 SW)>')
         self.assertEqual(im3.get_palette(), im.get_palette())
 
-        # It is still an error when the target format really does have
-        # an empty palette (all the entries are black).
-        self.assertRaises(pygame.error, im2.convert, 8)
-        self.assertEqual(pygame.get_error(), "Empty destination palette")
-
     def test_convert_init(self):
         """ Ensure initialization exceptions are raised
             for surf.convert()."""

